src/parquet/column/scanner-test.cc

```diff
@@ -0,0 +1,271 @@
+  vector<int16_t> rep_levels_;
+  vector<uint8_t> data_buffer_; // For BA and FLBA
+};
+
+template<>
+void TestFlatScanner<Type::INT96, Int96>::InitValues() {
+  random_Int96_numbers(num_values_, 0, std::numeric_limits<int32_t>::min(),
+      std::numeric_limits<int32_t>::max(), values_.data());
+}
+
+template<>
+void TestFlatScanner<Type::BYTE_ARRAY, ByteArray>::InitValues() {
+  int max_byte_array_len = 12;
+  int num_bytes = max_byte_array_len + sizeof(uint32_t);
+  size_t nbytes = num_values_ * num_bytes;
+  data_buffer_.resize(nbytes);
+  random_byte_array(num_values_, 0, data_buffer_.data(), values_.data(),
+      max_byte_array_len);
+}
+
+template<>
+void TestFlatScanner<Type::FIXED_LEN_BYTE_ARRAY, FLBA>::InitValues() {
+  size_t nbytes = num_values_ * FLBA_LENGTH;
+  data_buffer_.resize(nbytes);
+  random_fixed_byte_array(num_values_, 0, data_buffer_.data(), FLBA_LENGTH,
+      values_.data());
+}
+
+template<>
+void TestFlatScanner<Type::FIXED_LEN_BYTE_ARRAY, FLBA>::InitDescriptors(
+    std::shared_ptr<ColumnDescriptor>& d1, std::shared_ptr<ColumnDescriptor>& d2,
+    std::shared_ptr<ColumnDescriptor>& d3) {
+  NodePtr type = schema::PrimitiveNode::MakeFLBA("c1", Repetition::REQUIRED,
+      FLBA_LENGTH, LogicalType::UTF8);
+  d1.reset(new ColumnDescriptor(type, 0, 0));
+  type = schema::PrimitiveNode::MakeFLBA("c2", Repetition::OPTIONAL,
+      FLBA_LENGTH, LogicalType::UTF8);
+  d2.reset(new ColumnDescriptor(type, 4, 0));
+  type = schema::PrimitiveNode::MakeFLBA("c3", Repetition::REPEATED,
+      FLBA_LENGTH, LogicalType::UTF8);
+  d3.reset(new ColumnDescriptor(type, 4, 2));
+}
+
+
+typedef TestFlatScanner<Type::INT32, int32_t> TestFlatInt32Scanner;
+typedef TestFlatScanner<Type::INT64, int64_t> TestFlatInt64Scanner;
+typedef TestFlatScanner<Type::INT96, Int96> TestFlatInt96Scanner;
+typedef TestFlatScanner<Type::FLOAT, float> TestFlatFloatScanner;
+typedef TestFlatScanner<Type::DOUBLE, double> TestFlatDoubleScanner;
+typedef TestFlatScanner<Type::BYTE_ARRAY, ByteArray> TestFlatByteArrayScanner;
+typedef TestFlatScanner<Type::FIXED_LEN_BYTE_ARRAY, FLBA> TestFlatFLBAScanner;
+
+static int num_levels_per_page = 100;
+static int num_pages = 20;
+static int batch_size = 32;
+
+TEST_F(TestFlatInt32Scanner, TestScanner) {
+  ExecuteAll(num_pages, num_levels_per_page, batch_size);
+}
+
+TEST_F(TestFlatInt64Scanner, TestScanner) {
+  ExecuteAll(num_pages, num_levels_per_page, batch_size);
+}
+
+TEST_F(TestFlatInt96Scanner, TestScanner) {
+  ExecuteAll(num_pages, num_levels_per_page, batch_size);
+}
+
+TEST_F(TestFlatFloatScanner, TestScanner) {
+  ExecuteAll(num_pages, num_levels_per_page, batch_size);
+}
+
+TEST_F(TestFlatDoubleScanner, TestScanner) {
+  ExecuteAll(num_pages, num_levels_per_page, batch_size);
+}
+
+TEST_F(TestFlatByteArrayScanner, TestScanner) {
+  ExecuteAll(num_pages, num_levels_per_page, batch_size);
+}
+
+TEST_F(TestFlatFLBAScanner, TestScanner) {
+  ExecuteAll(num_pages, num_levels_per_page, batch_size);
+}
```

Comment:
I really don't think `TYPED_TEST` is all that scary but I'll let this slide for now.